packages/xod-cli/src/commands/publish.js

```diff
@@ -0,0 +1,165 @@
+    const password =
+      this.flags.password ||
+      (await cli.prompt(msg.passwordFlagDescription, { type: 'hide' }));
+
+    const targetUsername = onBehalf || username;
+
+    // load library
+    const { libname, version } = await pipeP(
+      () => this.cliActionStartPromise(msg.publishCommandLibLoading),
+      () => packLibVersion(projectPath),
+      this.tapCliActionStop()
+    )();
+    const libUri = createLibUri(targetUsername, libname, version.semver);
+
+    // get user's access token
+    const accessToken = await pipeP(
+      () => this.cliActionStartPromise(msg.publishCommandAuthentication),
+      () => getAccessToken(api, username, password),
+      this.tapCliActionStop()
+    )().catch(err =>
+      this.die(msg.publishCommandAuthError(err.statusText || err))
+    );
+
+    // check that user exists
+    await this.cliActionStartPromise(msg.publishCommandGetUser)
+      .then(() => getUser(api, targetUsername))
+      .catch(err => {
+        if (err.status === 404) {
+          this.die(msg.publishCommandUser404(targetUsername));
+        } else {
+          this.die(msg.publishCommandUserError(err));
+        }
+      })
+      // check access rights
+      .then(user => {
+        if (
+          user.username !== username &&
+          user.trusts.indexOf(username) === -1
+        ) {
+          return this.die(
+            msg.publishCommandAccessDenied(username, uriToString(libUri))
+          );
+        }
+        return user;
+      })
+      .then(this.tapCliActionStop());
+
+    const putLibrary = (user, name) =>
+      pipeP(
+        () => this.cliActionStartPromise(msg.publishCommandCreateLibrary),
+        () => putUserLib(api, accessToken, user, name),
+        this.tapCliActionStop()
+      );
+
+    // get library
+    await this.cliActionStartPromise(msg.publishCommandGetLib)
+      .then(() => getLib(api, targetUsername, libname))
+      .then(this.tapCliActionStop())
+      .catch(err => {
+        this.cliActionStop();
+        if (err.status !== 404) this.die(msg.publishCommandLibGetError(err));
+
+        // put library
+        return putLibrary(targetUsername, libname)().catch(err2 => {
+          if (err2.status === 403) {
+            this.die(
+              msg.publishCommandAccessDenied(username, uriToString(libUri))
+            );
+          } else {
+            this.die(msg.publishCommandLibPutError(err));
+          }
+        });
+      });
+
+    // post library
+    await pipeP(
+      () => this.cliActionStartPromise(msg.publishCommandLibPost),
+      () => postUserLib(api, accessToken, targetUsername, libname, version),
+      this.tapCliActionStop(),
+      () => exit(0)
+    )().catch(err => {
+      if (err.status === 409) {
+        this.die(msg.publishCommandVersionExists(uriToString(libUri)));
+      } else {
+        this.die(msg.publishCommandLibPostError(err));
+      }
+    });
+  }
+}
+
+PublishCommand.aliases = ['p'];
```

Comment:
Suggest avoiding aliases everywhere until some usage patterns will become apparent. Typing `publish` instead of `p` is not a big deal, but looks much better when shared in a discussion.